alephium/lib/deployment.ts

```diff
@@ -150,6 +155,49 @@ export class PrivateKeySigner extends SignerWithNodeProvider {
   }
 }
 
+async function isTxExists(provider: NodeProvider, txId: string): Promise<boolean> {
+  const txStatus = await provider.transactions.getTransactionsStatus({txId: txId})
+  return txStatus.type !== "TxNotFound"
+}
+
+function recordEqual(left: Record<string, string>, right: Record<string, string>): boolean {
+  const leftKeys = Object.keys(left)
+  const rightKeys = Object.keys(right)
+  if (leftKeys.length !== rightKeys.length) {
+    return false
+  }
+  for (const key of leftKeys) {
+    if (left[key] !== right[key]) {
+      return false
+    }
+  }
+  return true
+}
+
+async function needToDeploy(
```

Comment:
Great change :+1: 

Question: With this change (which makes it idempotent), do we still need to check the [lastFailedStep](https://github.com/alephium/wormhole-fork/blob/improve-deployment-workflow/alephium/lib/deployment.ts#L348)? If the user actually change e.g. the code after a failed deployment, does it pick it up?

Reply:
> With this change (which makes it idempotent), do we still need to check the [lastFailedStep](https://github.com/alephium/wormhole-fork/blob/improve-deployment-workflow/alephium/lib/deployment.ts#L348)?

I was also hesitant about this, if we check the `lastFailedSetp`, then we can skip the successfully executed scripts, which can avoid some extra checks.

> If the user actually change e.g. the code after a failed deployment, does it pick it up?

Yes, it will retry to deploy after a failed deployment regardless of whether the code changes, but if the deployment succeeds and the code changes, then it will skip re-deploy the contract, from this point of view, maybe we should skip to checking the `lastFailedStep`? WDYT?

Reply:
Yeah, it feels like if we expect that the deployment steps should be idempotent (and we have done our part to make it easy to do so) then it feels like we don't need to check the `lastFailedStep` any more. WDYT?

Reply:
Sounds great 👍 , I'll fix this.

Reply:
I found another problem, the script will fail if it is executed repeatedly, so we also need to skip the scripts which executed successfully before.

But because we use `Script.typeId` as the task key and users may run the script more than once([register_chain](https://github.com/alephium/wormhole-fork/pull/27/files#diff-4b96891847f33c9f9874491068010d6735e8918d0deba756a43f0c04b0cac88c) is an example), this means different tasks have same task key, so I add an extra optional params `taskName` to `Deployer.runScript`, users don't need this if one script is a task, I'm not sure if there is a better way to resolve this, do you have any suggestions?

Reply:
How about this approach:

For each script deployment, we add an optional tag to it and use `Script.typeId -> tag` as the task key. For your example, we will have `RegisterChain -> ETH`, `RegisterChain -> BSC` etc as key.

Reply:
Sounds great 👍